cmd/voucher/challenger/challenger.go

```diff
@@ -0,0 +1,99 @@
+// This Source Code Form is subject to the terms of the MIT License.
+// If a copy of the MIT License was not distributed with this
+// file, you can obtain one at https://opensource.org/licenses/MIT.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+package challenger
+
+import (
+	"bytes"
+	"context"
+	"crypto/rand"
+	"crypto/sha256"
+	"errors"
+	"os"
+
+	"github.com/dusk-network/dusk-blockchain/cmd/voucher/node"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/peer"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/message"
+	"github.com/dusk-network/dusk-blockchain/pkg/p2p/wire/topics"
+	log "github.com/sirupsen/logrus"
+)
+
+const challengeLength = 20
+
+// Challenger is the component responsible for vetting incoming connections.
+type Challenger struct {
+	nodes *node.Store
+}
+
+// New creates a new, initialized Challenger.
+func New(store *node.Store) *Challenger {
+	return &Challenger{nodes: store}
+}
+
+// SendChallenge to a connecting peer.
+func (c *Challenger) SendChallenge(ctx context.Context, r *peer.Reader, w *peer.Writer, ch chan bytes.Buffer) {
+	challenge, err := generateRandomBytes(challengeLength)
+	if err != nil {
+		log.Panic(err)
+	}
+
+	buf := bytes.NewBuffer(challenge)
+	if err := topics.Prepend(buf, topics.Challenge); err != nil {
+		log.Panic(err)
+	}
+
+	ch <- *buf
+
+	// Enter the node into the store, for future reference.
+	c.nodes.Add(r.Addr(), challenge)
```

Comment:
A kind of a race condition (very rare situation I believe), the voucher might receive `topics.Response` message before adding the peer to the registry thus the peer would be blacklisted. Shall we first add the peer and then send the wire message?

Reply:
It would be impossible, because the read loop isn't instantiated until the node is added to the list. The message might have arrived, but it won't be read yet.

Reply:
Makes sense!